Allow to specify runPolicy for BuildConfig

mfojtik opened the pull request:
* Add `runPolicy=Parallel | Serial | SerialLatestOnly` into the `BuildConfig`.
  * The `Parallel` represents current behavior.
  * The `Serial` means the builds will be execute sequentially.
  * The `SerialLatestOnly` means that only the latest build will be executed and previous non-running builds will be cancelled.

The `Serial` will be default when this PR gets merged. This will be documented for users.

mfojtik commented:
@bparees @smarterclayton this is an very early draft of the API additions needed for https://trello.com/c/Wz8FFwko/849-13-single-threaded-builds. Would appreciate quick review before I jump into actual implementation.

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -707,6 +739,12 @@ type BuildRequest struct {
+	// make the build created from this build requests be the only build in the
+	// queue. This option is valid only for build configurations with "recent"
+	// build scheduling policy.
+	SkipBuildsInQueue *bool
```
@smarterclayton @bparees maybe this deserves a better name ;-)

smarterclayton replied:
What does Jenkins call this flag?  How many flags do they have?  What about jenkins pipeline and concourse?

mfojtik replied:
@jstrachan && @jimmidyson to the rescue :-)

jimmidyson replied:
There isn't a direct equivalent afaik. There's also accelerated build which will run the new build immediately, but not cancel the other builds. It's all possible of course :) @jstrachan Do you know any different?

jstrachan replied:
not seen anything else equivalent I'm afraid

mfojtik replied:
@jimmidyson @jstrachan thanks guys, we are opting out this option and have one parallel and serial builds for now.

bparees replied:
not sure this is the way we want to expose that behavior anyway.  i think the use case is more like:

1) i started a build
2) i look at the build queue and see it's full of garbage
3) i run "oc skip-builds" (name tbd) which skips all the queued builds except the most recent one (mine).

mfojtik replied:
@bparees `oc cancel-build XYZ --skip range` ?

mfojtik replied:
@bparees iow.

`oc skip-build --to ruby-10`

bparees replied:
--to is fancy :)  but sure.

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -559,10 +566,35 @@ type BuildConfigSpec struct {
+	// they are created. Every build gets queued up and will execute when the
+	// previous build completes. Users can mark a build for an immediate execution
+	// using the SkipBuildsInQueue option on the build request.
+	BuildSchedulePolicyRecent BuildSchedulePolicy = "recent"
```
@smarterclayton @bparees not sure if "recent" is the right description of the behavior here. Also not 100% sure we even need this policy. Can't it just be the "serial" option? IOW. if I have "serial" policy, I can always clean up the queue by creating a build request that wipes out the queue.

bparees replied:
sounds like we're dumping this strategy for now, but i think the comment description is wrong anyway.

mfojtik replied:
going to opt-out this from the initial implementation, we can revisit this later.

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -142,6 +146,9 @@ const (
 	BuildPhaseCancelled BuildPhase = "Cancelled"
+
+	// BuildPhaseSkipped indicates that a queued build was skipped.
+	BuildPhaseSkipped BuildPhase = "Skipped"
```
@smarterclayton @bparees I think we should add new phase where (this is will termination phase) to indicate that this build was not cancelled but skipped (because user wiped out the queue or marked it to skip?)

kargakis replied:
@smarterclayton are we considering https://github.com/kubernetes/kubernetes/issues/7856 for OS resources?

smarterclayton replied:
Regarding 7856, yes we are considering it, but we won't remove phase from
any of our resources or now.  Be aware that you cannot add a new phase in a
backwards compatible way (it would break old clients that expect phases),
you can only add new conditions.  So this new change couldn't add a new
phase.

On Mon, Apr 11, 2016 at 8:06 AM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/build/api/types.go
> <https://github.com/openshift/origin/pull/8453#discussion_r59193077>:
>
> > @@ -142,6 +146,9 @@ const (
> >
> >  	// BuildPhaseCancelled indicates that a running/pending build was stopped from executing.
> >  	BuildPhaseCancelled BuildPhase = "Cancelled"
> > +
> > +	// BuildPhaseSkipped indicates that a queued build was skipped.
> > +	BuildPhaseSkipped BuildPhase = "Skipped"
>
> @smarterclayton <https://github.com/smarterclayton> are we considering
> kubernetes/kubernetes#7856
> <https://github.com/kubernetes/kubernetes/issues/7856> for OS resources?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/0377afaa8ee7bb504a6e7484892b1dbeb5be3e41#r59193077>
>

mfojtik replied:
@smarterclayton k, so I will proceed with 'canceled' instead of adding new phase... perhaps we can describe the reason of build being canceled in annotation. Or introduce new field "skipped" that will be set to true if the build was skipped (phase will be cancelled).

mfojtik replied:
@smarterclayton note that we already have field "cancelled" ;-)

bparees replied:
I guess this raises a question of what the manual action is going to be.  I thought it would be "delete all the queued builds" in which case we don't need a new phase.  But if the manual action will be "skip all the queued builds except the newest one, and run the newest one" then it's debatable.  There's still an argument you could just delete all the builds that weren't run, but Skipped is probably better for audit purposes.

so in conclusion yes i think we want this, but you'll need to be careful because there are a lot of places in code where we look for builds in a "terminated" state and we need to make sure they now treat this as a terminated state.  I think we have an array somewhere of all the "terminal" states that you can just add this to, but it may not be used consistently.  i'd audit the code for anywhere we reference any of the existing terminal state constants.

bparees commented on pkg/build/api/types.go:
```diff
@@ -86,6 +86,10 @@ type BuildStatus struct {
 
+	// Position represents the build position in the build queue. Only applies to
+	// build configurations with "serial" or "recent" scheduling policy.
+	Position *int64
```
going to use the build number annotation instead?

mfojtik replied:
yes, already updated this pull.

mfojtik commented:
@bparees @smarterclayton PTAL. I did some initial implementation for serial builds. 

It works, not it is not ideal.. I mean, the builds are executed sequentially based on the build number we store in the annotation.
The problem is that the time for retrying the new build handing seems very long. When one build finishes, it took about ~30 seconds when the next build is scheduled. I dunno if this is something we can tweak (build completion force trigger handing builds in New state).

Also the BuildController has 30 minutes retry timeout... does that mean if I have big queue for builds and they will not finish in 30 minutes, they will be marked as failed? Another solution would be to have separate controller that handle only "serial" builds.

mfojtik commented:
@smarterclayton @bparees little demo: https://www.youtube.com/watch?v=AEEOgmjnNHQ

bparees commented:
> When one build finishes, it took about ~30 seconds when the next build is scheduled. I dunno if this is something we can tweak (build completion force trigger handing builds in New state).

that should definitely be done.  I guess i didn't put it in the card but i was assuming that controller should, on seeing a build enter a terminal state, revisit the queue and pick the next build.

doesn't solve the "stuck build that is waiting for an IST" scenario though.

mfojtik commented:
@bparees it basically means that an event of "complete" build should kick HandleBuild() for all queued builds....

smarterclayton commented on api/swagger-spec/oapi-v1.json:
```diff
@@ -17354,6 +17354,10 @@
      },
+     "schedulePolicy": {
+      "$ref": "v1.BuildSchedulePolicy",
+      "description": "SchedulePolicy describes how the new build created from this build configuration will be scheduled for execution. This is optional, if not specified we default to \"parallel\"."
```
I think we would default to serial?

mfojtik replied:
@smarterclayton right. I was thinking about that and I think it make sense. Serial is also default in Jenkins.

bparees replied:
i was thinking we'd default to parallel for backwards compatibility reasons.

i'm ok w/ changing it, but we need to acknowledge we're changing the default behavior for users.

mfojtik replied:
@smarterclayton i found the `schedulePolicy` long to type ;-) WDYT about changing that just to `schedule: parallel` ?

mfojtik replied:
@smarterclayton should we still default to Serial or SerialLatestOnly? I think Serial will be "safe" default.

smarterclayton commented on pkg/build/api/types.go:
```diff
@@ -86,6 +95,10 @@ type BuildStatus struct {
 	// Cancelled describes if a cancel event was triggered for the build.
 	Cancelled bool
 
+	// Skipped indicates that this build was marked as skipped. The build will be
```
This should be a condition, not a boolean.  Add a conditions array, and we need to define a condition that represents "not yet ready to be scheduled" (it's similar to pending or waiting or queued).  We need to tighten up terminology - I don't know that I like `schedule` as the verb here because that already means something specific elsewhere.  On Jobs we use `Parallelism`, but do we really anticipate multiple policies?  Can you describe a 3rd policy?

bparees replied:
skipped is not the same as "not yet ready".  skipped represents a final state.

arguably it *should* be a build state (like completed/failed/etc).  or perhaps we don't need it at all and we can just have the state be "canceled" and the reason would be "skipped due to newer build request"

mfojtik replied:
@smarterclayton @bparees i'm going to remove this for now and we can revisit "skipped" builds later when we implement the auto-skip logic. Right now, users will have to cancel builds in the queue manually using `oc cancel-build`. We discussed with @bparees on IRC that it would be nice to provide a 'range' option to `cancel-build` so you don't have to list all builds... Other option is to add something like: `oc cancel-build BCNAME --all-queued` which will cancel all builds that are not running. Basically cleanup the queue. WDYT?

bparees replied:
@mfojtik We also discussed in the card(I think?) having a label on builds that are queued so you could do a delete by label to delete all queued builds...

mfojtik replied:
@bparees hmm that makes sense :-) i can use that label instead of start timestamp.

smarterclayton commented on pkg/build/api/types.go:
```diff
@@ -177,6 +190,9 @@ const (
 	StatusReasonExceededRetryTimeout = "ExceededRetryTimeout"
+
+	// StatusReasonSkipped indicates that the build was skipped.
+	StatusReasonSkipped = "Skipped"
```
The reason would not be the same as the condition - the reason would be *why*, like "BuildRunning" or "WaitingForBuild"

bparees replied:
for now the only reason a build will be skipped is that the user took a manual action to skip over all the queued builds in order to run the most recent build request.

mfojtik replied:
technically "skipped" means cancelled, but we want to leave some notion there for auditing... also the cancel-build will have to have some option to set that.

bparees commented:
> @bparees it basically means that an event of "complete" build should kick HandleBuild() for all queued builds....

not just complete.  also failed, canceled, etc.

mfojtik commented:
@bparees made some improvements in the last commit, https://www.youtube.com/watch?v=0MpmnYebdHw

smarterclayton commented:
I think you guys need to define what skipped is before we argue about how
it's represented in the API :)

On Tue, Apr 12, 2016 at 3:44 PM, Michal Fojtik <notifications@github.com>
wrote:

> @bparees <https://github.com/bparees> made some improvements in the last
> commit, https://www.youtube.com/watch?v=0MpmnYebdHw
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453#issuecomment-209073682>
>

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +82,60 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+		if nextBuild != nil {
+			glog.Infof("Queue: The next build is %s", nextBuild.Name)
+			switch build.Status.Phase {
+			case buildapi.BuildPhaseError, buildapi.BuildPhaseFailed, buildapi.BuildPhaseComplete, buildapi.BuildPhaseCancelled:
```
pkg/build/util/util.go:IsBuildComplete

bparees commented:
> I think you guys need to define what skipped is before we argue about how it's represented in the API :)

to me, skipped is a reason.  canceled is the state.

canceled - because the user canceled it
canceled - because it was auto-skipped due to the presence of a newer build (function we are not implementing yet)
canceled - because the user requested we skip queued builds and run the most recent one instead

mfojtik commented:
@bparees +1 we don't want to introduce new phase, so we have to live with existing phases. The best fit for "skipped" is cancelled with a good reason (Ben described the reasons).

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -559,10 +556,30 @@ type BuildConfigSpec struct {
+	// SchedulePolicy describes how the new build created from this build
+	// configuration will be scheduled for execution.
+	// This is optional, if not specified we default to "parallel".
+	SchedulePolicy *BuildSchedulePolicy
```
@smarterclayton API question... do you think this field should be here or in BuildSpec? I think it should be higher level (same as Triggers), WDYT?

bparees replied:
Since this doesn't apply to Builds, it makes sense to me that it is not part of BuildSpec.

mfojtik replied:
right. I think it should live on BuildConfig as it is now.

bparees replied:
but renamed i guess per the type renaming below.

mfojtik replied:
@bparees yes as soon as we decide the name :-)

mfojtik commented:
[testonlyextended][extended:core(builds are executed in sequence)]

mfojtik commented:
@bparees @smarterclayton I guess this is ready for formal API/code review now.

jhadvig commented:
@jwforres with this landing, we might rework our `Start Build` button logic since by default it seems that the builds will be scheduled in series, so there wont be any point in disabling the button.

mfojtik commented:
@bparees the extended test flakes are mystery to me, I cannot reproduce them locally... also the first extended test run for this PR succeeded...

jwforres commented:
Good to hear, would love to take out the disabling.  @jhadvig go ahead and
open a github issue to track removing that
On Apr 14, 2016 6:27 AM, "Jakub Hadvig" <notifications@github.com> wrote:

> @jwforres <https://github.com/jwforres> with this landing, we might
> rework our Start Build button logic since by default it seems that the
> builds will be scheduled in series, so there wont be any point in disabling
> the button.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453#issuecomment-209868809>
>

mfojtik commented:
@jwforres I think the UI impact (except disabling the disabling the start build button ;-) is that the BuildConfig will most likely get a new field with `schedulingPolicy` we should show in the BuildConfig details.

jhadvig commented:
@mfojtik @jwforres also maybe an option to change the `schedulingPolicy` in the BC editor ?

mfojtik commented:
@jhadvig if you change that property while you have builds queued, it does not make the builds execute in parallel immediately, but new builds are scheduled immediately... So changing that property should come with a warning to user ;-)

jwforres commented:
Is this just a simple option to either be Serial or Parallel?  If thats all
it is then adding it to the editor wouldn't be too bad.

On Thu, Apr 14, 2016 at 8:15 AM, Jakub Hadvig <notifications@github.com>
wrote:

> @mfojtik <https://github.com/mfojtik> @jwforres
> <https://github.com/jwforres> also maybe an option to change the
> schedulingPolicy in the BC editor ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453#issuecomment-209910126>
>

mfojtik commented:
@jwforres yes, two options where "Serial" will be most likely the new default.

kargakis commented on pkg/build/util/util.go:
```diff
@@ -72,6 +73,32 @@ func IsPaused(bc *buildapi.BuildConfig) bool {
+}
+
+// BuildConfigName returns the build config associated with the build.
+func BuildConfigName(build *buildapi.Build) string {
```
ConfigNameForBuild?  https://github.com/openshift/origin/pull/8453/files#diff-388ced36fbdce0a3359ed094e4889609L95

mfojtik replied:
right :-) i'm getting blind.

kargakis commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
-			return fmt.Errorf("Failed to cancel build %s/%s: %v, will retry", build.Namespace, build.Name, err)
+	if build.Status.Phase != buildapi.BuildPhaseCancelled {
+		var reason buildapi.StatusReason
+		switch {
```
use a simple if?

mfojtik replied:
@kargakis I kind of like switch because it makes it better to read ;-) is there any pragmatic reason to use if here?

bparees commented:
> if you change that property while you have builds queued, it does not make the builds execute in parallel immediately, but new builds are scheduled immediately... So changing that property should come with a warning to user

@mfojtik if you change that setting, the next time we go to pick a build to run (either because a new build comes in and we look at it, or because an existing build finishes and we go to pick the next one, or because an existing build gets updated so we get an event for it), we should start all the builds in the queue.

mfojtik commented:
@bparees that will mean we have to monitor the BC (have to query BC every time we pick a next build to run). I think that all builds created from the "serial" BC should be executed in "serial" order and changing the BC to "parallel" should not change them, but should affect every "new" build created from that BC. Agree?

bparees commented:
> @bparees that will mean we have to monitor the BC (have to query BC every time we pick a next build to run). I think that all builds created from the "serial" BC should be executed in "serial" order and changing the BC to "parallel" should not change them, but should affect every "new" build created from that BC. Agree?

ostensibly I agree, but i'm not sure i have my head wrapped around what the behavior will be when someone stacks up a build queue like:

SSSSPPPPSSSSPPPPSSSSPPPP

?

which ones will run serially, which will run in parallel, and when?

bparees commented:
(I don't think this is anything other than an edge case, so i don't want to spend too much time debating it or writing complex handling for it, but we should at least have a sense of the implications)

mfojtik commented:
@bparees the builds created from "serial" BC will run as serial, the builds created from "parallel" BC will run immediately (IOW. they will skip the queue, and run in parallel with the currently running serial build).

bparees commented:
So...serially builds aren't?  that doesn't sound ideal :)

mfojtik commented:
@bparees so here are some scenarios:

1) Create "serial" BC, start 5 builds. The order will be: B1, B2, B3, B4, B5.
2) Change the BC to "parallel" when the queue is executing B3. Kick new build. That build will execute right after it is created (parallel with B3). The B4 will wait for that build to complete.
3) Create "parallel" BC, start 5 builds. The order will be random (all builds are scheduled ASAP). Then during the builds are still running, switch the BC to "serial". Then kick 2 builds. Now what will happen is that those builds will wait till all parallel builds are complete and then they will execute...

liggitt commented:
so the gate is just on starting a serial build, and not on preventing other builds from running parallel with a serial build?

mfojtik commented:
@liggitt basically yes... the builds are marked to run in sequence when they are created from BC... we don't change that when you change the BC on the fly.

mfojtik commented:
@liggitt if you have serial builds queued up, they will never be scheduled to run if there is another build running already... so if you change the BC to parallel, kick some builds, the serial builds that were created will hold until the parallel builds all complete and then move on... I think that is acceptable and we really talking about an edge case (I don't think users will go update BC scheduling policy regularly ;-)

liggitt commented:
seems like there's a third option lurking in there... for lack of a better term, "serial with deduping". Here's what I expect to be a common scenario:

1. BuildConfig set up with serially with webhooks so source pushes trigger builds
2. I commit and push rapidly, which queues up builds
3. Typically, I only care about the most recent queued build... I wouldn't want to wait for 5 builds to run to get to the build for my latest commit.

Do we have an idea how we would dedupe the pending queue, even if we don't do that work now?

mfojtik commented:
@liggitt yes, we have follow up card for "auto-skipping" builds... that means you should be able to skip the queue by cancelling all queued builds and execute only last build... will that fit to your scenario?

(it might be a build request option?)

liggitt commented:
I would be annoyed if my workflow was "commit, push, commit, push, commit, push, cancel cancel cancel" :)

mfojtik commented:
@liggitt right but that applies to a flow when you do these rapidly :-) as I said, we should help you do this rapidly with auto-skipping.

bparees commented:
> I would be annoyed if my workflow was "commit, push, commit, push, commit, push, cancel cancel cancel" :)

it wouldn't be cancel cancel cancel, it would be a single "skip" to skip to the end, canceling everything in between automatically.

And that's really no different than jenkins behavior today (except jenkins doesn't give you an easy "cancel everything except the most recent request" option).

Developers will have to choose if they want this or not.  It's more critical for merge pipelines or production flows.  For an individual developer, they might choose to run parallel.  Or they might use an rsync flow to update instead of doing a new build each time.

liggitt commented:
I thought parallel builds weren't likely to work well (or it was ambiguous what the resulting image would be if they raced to finish). Serial is a great improvement, it just seemed like a variant of serial that collapsed pending builds would be a common need, but maybe I'm projecting my own workflow too much.

bparees commented:
> I thought parallel builds weren't likely to work well (or it was ambiguous what the resulting image would be if they raced to finish).

correct, that's the reason behind doing this feature.  

We originally considered a third mode of operation which was to always skip to the most recent.  That is, whenever there is a build queue, when the previous build finishes, skip all builds in the queue except the most recent (cancel them) and immediately run most recent item, which I think is what you're lobbying for.

At this point I think that is a follow on (@mfojtik that should be covered in your follow on card though, if it's not)

liggitt commented:
> whenever there is a build queue, when the previous build finishes, skip all builds in the queue except the most recent (cancel them) and immediately run most recent item, which I think is what you're lobbying for.

yeah, whether that logic is applied at build create time (to avoid lots of pending builds) or build start time (to skip pending builds)

> At this point I think that is a follow on (@mfojtik that should be covered in your follow on card though, if it's not)

totally agree, just wanted to mention it in case it affected anything being done here

bparees commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -367,6 +367,7 @@ func describeCustomStrategy(s *buildapi.CustomBuildStrategy, out *tabwriter.Writ
 // DescribeTriggers generates information about the triggers associated with a buildconfig
 func (d *BuildConfigDescriber) DescribeTriggers(bc *buildapi.BuildConfig, out *tabwriter.Writer) {
 	describeBuildTriggers(bc.Spec.Triggers, out)
+	formatString(out, "\nBuilds Schedule", buildapi.SchedulePolicyType(bc.Spec.SchedulePolicy))
```
won't string(bc.Spec.SchedulePolicy) work here instead of creating the Type function?

bparees replied:
also this doesn't belong in DescribeTriggers.

mfojtik replied:
@bparees it will but not if the SchedulePolicy is `nil`, so I need a function to set the default.

mfojtik replied:
and yes, this should be somewhere else :-)

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+		}
+		if nextBuild != nil {
+			glog.V(5).Infof("Build configuration %s/%s is serial, next build is %s", build.Namespace, bcName, nextBuild.Name)
+			if buildutil.IsBuildComplete(build) {
```
can't you check this a lot earlier in the logic?

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+	// If we receive build that reached its final state and the build was created
+	// from a serial build configuration, tag the next build in the queue for
+	// execution. That will produce event and the build should get executed
+	// immediately.
```
don't you also need to do this if the build that just completed was a parallel build (as long as no other parallel builds are running).

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+		case build.Status.Cancelled:
+			reason = buildapi.StatusReasonCancelBuildFailed
+		}
+		if len(reason) != 0 {
```
why isn't the content of this "if" just part of the case statement above?

mfojtik replied:
leftover from removing the skipped state...

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+			glog.V(5).Infof("Setting reason for build %s/%s to %q", build.Namespace, build.Name, reason)
+			if err := bc.CancelBuild(build); err != nil {
+				return fmt.Errorf("Failed to cancel build %s/%s: %v, will retry", build.Namespace, build.Name, err)
+			}
```
i don't see why this entire block was rewritten.  the only net change i see is you added a log statement.

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+		}
+		if nextBuild.Name != build.Name {
+			return nil
+		}
```
collapse to single if.

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -108,6 +163,44 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
+			nextBuild = &builds.Items[i]
+			previousBuildNumber = buildNumber
+			continue
+		}
```
if you initialize previousBuildNumber to "-1" you don't need this special case check (the first buildNumber you see will be guaranteed to be < previousBuildNumber)

bparees replied:
(in fact even initializing it to zero is sufficient since we start our build numbers at 1 i believe)

mfojtik replied:
yes, we start with 1, I will rework this :-)

smarterclayton replied:
I'll comment here (without digging further) that Jordan's scenario is
probably the one I would expect to see most often in practice on
OpenShift.  I don't think most people care if 3 triggers fire and result in
the last build.

On Sun, Apr 17, 2016 at 4:12 PM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/controller/controller.go
> <https://github.com/openshift/origin/pull/8453#discussion_r59991583>:
>
> > +	}
> > +	var (
> > +		nextBuild           *buildapi.Build
> > +		previousBuildNumber int64
> > +	)
> > +	for i, b := range builds.Items {
> > +		buildNumber, err := buildutil.BuildNumber(&b)
> > +		if err != nil {
> > +			glog.Errorf("Error getting build number: %v", err)
> > +			continue
> > +		}
> > +		if previousBuildNumber == 0 {
> > +			nextBuild = &builds.Items[i]
> > +			previousBuildNumber = buildNumber
> > +			continue
> > +		}
>
> yes, we start with 1, I will rework this :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/68d0bdac5916b17673695e06451439bfa55941b9#r59991583>
>

mfojtik replied:
@smarterclayton the CI/CD executors might care about processing all created builds? This comes down whether we
should kick a new build if there is already a build with the same inputs. The problem with that sort of validation is
that the builder image state is not predictable (you might update the builder image between scheduled builds).

I agree that the "rapid" option, where every trigger will create a build that will get scheduled next and cancelling all
the builds in the queue is valuable (and mostly like all users will use that for development). Maybe we need three
scheduling policies:

* **parallel** - current behavior, every build gets executed immediately after creation, no queue.
* **serial** - mimic the Jenkins build scheduling, builds are queued and executed in the serial order
* **rapid** - (I struggle to give this a name ;-) - same as serial, but without queue, the last build created is always executed as next, cancelling the previous build.

@bparees @liggitt what do you think?

mfojtik replied:
@bparees one concern with having three policies I have is that the switching of policies on-the-fly can cause a mess :)

smarterclayton replied:
Put the strategy behind an interface.

On Mon, Apr 18, 2016 at 6:43 AM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/controller/controller.go
> <https://github.com/openshift/origin/pull/8453#discussion_r60037780>:
>
> > +	}
> > +	var (
> > +		nextBuild           *buildapi.Build
> > +		previousBuildNumber int64
> > +	)
> > +	for i, b := range builds.Items {
> > +		buildNumber, err := buildutil.BuildNumber(&b)
> > +		if err != nil {
> > +			glog.Errorf("Error getting build number: %v", err)
> > +			continue
> > +		}
> > +		if previousBuildNumber == 0 {
> > +			nextBuild = &builds.Items[i]
> > +			previousBuildNumber = buildNumber
> > +			continue
> > +		}
>
> @bparees <https://github.com/bparees> one concern with having three
> policies I have is that the switching of policies on-the-fly can cause a
> mess :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/68d0bdac5916b17673695e06451439bfa55941b9#r60037780>
>

smarterclayton replied:
On Mon, Apr 18, 2016 at 6:40 AM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/controller/controller.go
> <https://github.com/openshift/origin/pull/8453#discussion_r60037443>:
>
> > +	}
> > +	var (
> > +		nextBuild           *buildapi.Build
> > +		previousBuildNumber int64
> > +	)
> > +	for i, b := range builds.Items {
> > +		buildNumber, err := buildutil.BuildNumber(&b)
> > +		if err != nil {
> > +			glog.Errorf("Error getting build number: %v", err)
> > +			continue
> > +		}
> > +		if previousBuildNumber == 0 {
> > +			nextBuild = &builds.Items[i]
> > +			previousBuildNumber = buildNumber
> > +			continue
> > +		}
>
> @smarterclayton <https://github.com/smarterclayton> the CI/CD executors
> might care about processing all created builds? This comes down whether we
> should kick a new build if there is already a build with the same inputs.
> The problem with that sort of validation is
> that the builder image state is not predictable (you might update the
> builder image between scheduled builds).
>

CI might - but not non-CI


> I agree that the "rapid" option, where every trigger will create a build
> that will get scheduled next and cancelling all
> the builds in the queue is valuable (and mostly like all users will use
> that for development). Maybe we need three
> scheduling policies:
>
>    - *parallel* - current behavior, every build gets executed immediately
>    after creation, no queue.
>    - *serial* - mimic the Jenkins build scheduling, builds are queued and
>    executed in the serial order
>    - *rapid* - (I struggle to give this a name ;-) - same as serial, but
>    without queue, the last build created is always executed as next,
>    cancelling the previous build.
>
>
"latest"


>
>
> @bparees <https://github.com/bparees> @liggitt
> <https://github.com/liggitt> what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/68d0bdac5916b17673695e06451439bfa55941b9#r60037443>
>

bparees commented on test/extended/builds/serial.go:
```diff
@@ -0,0 +1,86 @@
+			}
+
+			// Expect the builds are started in same order as they were created
+			o.Expect(startSequence).Should(o.BeEquivalentTo(executionSequence))
```
it would be harder, but better, if you actually created the builds out of order (you'd have to manually create the build objects w/ the build number annotation pre-set).

create build 1, wait until build1 is running, then create build 3, then build 2.  you should still see them execute/complete as 1,2,3.

mfojtik replied:
@bparees having the interface for policies means that I can make some unit tests for each policy and also HandleBuild()... so maybe this extended test will not be needed if the unit tests proves that the scheduling works as expected.

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +81,74 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
+	// immediately.
+	var nextBuild *buildapi.Build
+	schedulePolicy := buildutil.BuildSchedulePolicy(build)
+	if schedulePolicy == buildapi.BuildSchedulePolicySerial {
```
i'm pretty sure this logic is broken for any case where you have parallel builds completing and serial builds in the queue.

mfojtik replied:
retry controller runs for every build, including the serial builds... that means if the queue is free (no builds are running) retry controller will call HandleBuild on the each serial build... if it hits the first serial build, it will be executed :-) It will just not be that fast I guess because you will have to wait till retry controller kicks in (which might take couple seconds)...

I think that is fine for now?

mfojtik commented:
@smarterclayton @liggitt @bparees 

I added third policy and tested it (but I want to spend more time testing it under load). The policy name is `rapid`, but please suggest me a better name ;-)

Small demo: https://youtu.be/_bxA7WD-aE0 

(note that in demo a "running" build was cancelled, that is fixed now as we should not cancel a running builds).

mfojtik commented:
@bparees also addressed hopefully all your comments (thanks!)

bparees commented:
@mfojtik name ideas:
serial_skip
skip_intermediate
serial_skip_intermediate

bparees commented:
> retry controller runs for every build, including the serial builds... that means if the queue is free (no builds are running) retry controller will call HandleBuild on the each serial build (it is in fact retrying them all the time ;-)... if it hits the first serial build, it will be executed :-) It will just not be that fast I guess because you will have to wait till retry controller kicks in (which might take couple seconds)...

@mfojtik what are you calling the retry controller?  are you referring to the reget/list behavior that occurs every 2 minutes?  because we should not be relying on that to handle these scenarios.

mfojtik commented:
@bparees I think this is now fixed (we check every build and when it complete, we get the next build and handle it). in case you have parallel build finishing, it will call getNextBuild and notify the first serial build in the queue.

mfojtik commented:
@bparees how about naming it "latest" (BuildSchedulePolicy=latest)... It is always executing the "latest" build.

mfojtik commented:
FYI. "latest-only" it is :-)

smarterclayton commented:
Remember that these need to follow Kube conventions, so it'll be Parallel, Serial, and LatestOnly

smarterclayton commented:
actually - if Serial means in order then this should be SerialLatestOnly

liggitt commented:
or, if you want a serial build that cancels any earlier pending builds, you could call it `SerialKiller`

bparees commented:
@liggitt i award you one internet.

mfojtik commented:
@liggitt rofl :)

mfojtik commented:
@bparees @smarterclayton I moved the scheduling strategies and logic to a separate interface, renamed "rapid" to "latest-only", etc. PTAL

(still require a lot of tests, but want to get another round of review to spot naming issues :-)

smarterclayton commented on pkg/build/api/types.go:
```diff
@@ -559,10 +558,35 @@ type BuildConfigSpec struct {
+const (
+	// BuildSchedulePolicyParallel schedules new builds immediately after they are
+	// created. Builds will be executed in parallel. This is the default policy.
+	BuildSchedulePolicyParallel BuildSchedulePolicy = "parallel"
```
Parallel

smarterclayton commented on pkg/build/api/types.go:
```diff
@@ -559,10 +558,35 @@ type BuildConfigSpec struct {
+	// BuildSchedulePolicySerial schedules new builds to execute in a sequence as
+	// they are created. Every build gets queued up and will execute when the
+	// previous build completes.
+	BuildSchedulePolicySerial BuildSchedulePolicy = "serial"
```
`Serial`

smarterclayton commented on pkg/build/api/types.go:
```diff
@@ -559,10 +558,35 @@ type BuildConfigSpec struct {
+	// BuildSchedulePolicyRapid schedules new builds to execute in a sequence but
+	// cancelling previous builds if there are any. This allows newly triggered
+	// build to be always executed as "next".
+	BuildSchedulePolicyRapid BuildSchedulePolicy = "rapid"
```
Unless we have a really good reason, `SerialMostRecent` or `SerialLatest` or whatever, vs something that doesn't have the `Serial` prefix.

smarterclayton commented on pkg/build/api/util.go:
```diff
@@ -24,6 +24,15 @@ func StrategyType(strategy BuildStrategy) string {
+// BuildSchedulePolicy.
+func SchedulePolicyType(policy *BuildSchedulePolicy) string {
+	if policy == nil {
+		return "serial"
```
This is not where we should be doing defaulting - that should always be in the API.

bparees replied:
in which case this whole function can go away.

smarterclayton commented on pkg/build/api/v1/types.go:
```diff
@@ -525,6 +525,11 @@ type BuildConfigSpec struct {
+	// SchedulePolicy describes how the new build created from this build
+	// configuration will be scheduled for execution.
+	// This is optional, if not specified we default to "parallel".
+	SchedulePolicy *BuildSchedulePolicy `json:"schedulePolicy,omitempty"`
```
You can't use `schedule` for this, it's going to be confused with scheduling.  It has to be `TriggerPolicy` or similar (since it's based on triggers).

bparees replied:
ExecutePolicy?  i don't like TriggerPolicy as we already have that for our list of triggers.

bparees replied:
OrderingPolicy?

mfojtik replied:
NewBuildPolicy? StartPolicy? (I like the later, because it describes the purpose of this field... it is a policy of how the builds are started - serially, parallel or latest build only).

bparees commented on pkg/build/api/types.go:
```diff
@@ -559,10 +558,34 @@ type BuildConfigSpec struct {
+
+	// BuildSchedulePolicySerialLatestOnly schedules only the latest build to execute,
+	// cancelling all the previously queued build.
+	BuildSchedulePolicySerialLatestOnly BuildSchedulePolicy = "latest-only"
```
SerialLatestOnly

bparees commented on pkg/build/api/validation/validation.go:
```diff
@@ -82,6 +82,15 @@ func ValidateBuildConfig(config *buildapi.BuildConfig) field.ErrorList {
+		case buildapi.BuildSchedulePolicyParallel, buildapi.BuildSchedulePolicySerial, buildapi.BuildSchedulePolicySerialLatestOnly:
+		default:
+			allErrs = append(allErrs, field.Invalid(specPath.Child("schedulePolicy"), *schedulePolicy, "schedule policy must parallel, latest-only, or serial"))
+		}
```
this will need to be updated to force schedulePolicy to be non-nil once we're defaulting it in the api.

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -79,21 +83,53 @@ func (bc *BuildController) CancelBuild(build *buildapi.Build) error {
 	// A cancelling event was triggered for the build, delete its pod and update build status.
-	if build.Status.Cancelled && build.Status.Phase != buildapi.BuildPhaseCancelled {
+	if build.Status.Phase != buildapi.BuildPhaseCancelled && build.Status.Cancelled {
+		build.Status.Reason = buildapi.StatusReasonCancelBuildFailed
```
why are you setting this here?  it should only be set if the cancel attempt failed.

bparees commented on pkg/build/controller/scheduler/scheduler.go:
```diff
@@ -0,0 +1,143 @@
+		return false, nil
+	}
+	glog.V(5).Infof("Build %s/%s scheduler is latest-only", build.Namespace, build.Name)
+	nextBuild, err := GetNextBuild(s.OSClient.Builds(build.Namespace), build)
```
not clear to me how this works.  Isn't GetNextBuild still going to return the next build in the list, not the most recent build in the list?

mfojtik replied:
yeah, this logic is a little bit more complicated that I want...

This will run for every "latest-only" build (and I assume in random order ;-).
When the current build **is** the next build (mean it should run next in serial order) it will cancel the build(s) before it using [this](https://github.com/openshift/origin/pull/8453/files/31c441e199e9c4037b91dd540762313dcf7d1ffd#diff-d66f21265a02ba7183dc3c960bfe9902R75) logic. Using this, it will cancel everything except the "last" build that will be next :-)

As I said, it is more complicated and I should make it more clear... Problem is you have to deal with the fact that HandleBuild runs in non-predictive way ;-)

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -416,6 +416,20 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
+		case buildapi.BuildSchedulePolicySerial:
+			build.Labels[buildapi.BuildQueuedLabel] = "true"
+		case buildapi.BuildSchedulePolicySerialLatestOnly:
+			build.Labels[buildapi.BuildQueuedLabel] = "false"
```
debatable whether this should have the queued label..  certainly when you have just 2 builds (one running, one next) it makes sense to call that 2nd build "queued".

when you have 3, so the middle one will never run, calling it queued is more debatable.

mfojtik replied:
@bparees the "false" is confusing here but I don't want to introduce another label for "latest-only" builds... What I tried to say here is that when the build is latest-only, it will not be queued, but executed next (queued=false). The middle one will never run if you create third one, but until you create the third one, it is the one that will run next...

bparees replied:
what benefit does the label offer over queued=true? do we need to introduce the distinction?

bparees commented on pkg/build/util/util.go:
```diff
@@ -72,6 +73,30 @@ func IsPaused(bc *buildapi.BuildConfig) bool {
+			return buildapi.BuildSchedulePolicySerial
+		}
+	}
+	return buildapi.BuildSchedulePolicyParallel
```
this is weird.  you've created a distinction between queued label=false and queued label not being present.  that doesn't seem very intuitive.

mfojtik replied:
@bparees agree, but the assumption was that users will never set this label manually ;-) Do you suggest a new label for latest-only?

mfojtik replied:
@bparees actually I'm thinking about something like "io.openshift/build-start-policy=<PolicyName>"...

bparees replied:
what are we going to use the label for?

mfojtik replied:
@bparees right now, only to set what policy BC had when we created this build...

bparees commented on pkg/build/controller/scheduler/scheduler.go:
```diff
@@ -0,0 +1,143 @@
+		return false, nil
+	}
+	glog.V(5).Infof("Build %s/%s scheduler is parallel", build.Namespace, build.Name)
+	return true, nil
```
this could be slightly fancier and return false if any serial build is currently running.

mfojtik replied:
@bparees so you won't be able to schedule any parallel build till you have running serial build? when I change my BC to parallel, I expect new builds to run immediately in parallel with whatever is running currently :-)

bparees replied:
> so you won't be able to schedule any parallel build till you have running serial build?  

?

you mean "you won't be able to run any parallel builds while you have a running serial build?"  If so then yes, that's what i mean.

> when I change my BC to parallel, I expect new builds to run immediately in parallel with whatever is running currently :-)

when i change my BC to parallel, i expect builds i ran as serial builds to continue to run in isolation.

mfojtik commented:
@bparees bumped, i took a video of Serial vs. SerialLatestOnly: https://www.youtube.com/watch?v=ivpVBznIFS8

mfojtik commented:
@bparees also moved handing Complete() event into scheduler so different strategies can have own handlers for "complete" build.

mfojtik commented:
@smarterclayton we agreed on `StartPolicy` as a field name, does it work for you?

mfojtik commented:
[test]

mfojtik commented on pkg/build/controller/scheduler/scheduler.go:
```diff
@@ -0,0 +1,225 @@
+package scheduler
```
this is under refactoring now :-)

mfojtik commented:
@smarterclayton @bparees hopefully last refactoring before I start crafting unit tests. The changes I made:

* `s/schedulePolicy/runPolicy/`
* Get rid of "pkg/build/controller/schedule" and split all run policies into files under "pkg/build/controller/policy" package. (each file will get its own unit test).
* A lot of small fixes / de-duplication of errors, etc.

PTAL.

smarterclayton commented:
I love me some videos

On Tue, Apr 19, 2016 at 6:56 AM, Michal Fojtik <notifications@github.com>
wrote:

> @bparees <https://github.com/bparees> bumped, i took a video of Serial
> vs. SerialLatestOnly: https://www.youtube.com/watch?v=ivpVBznIFS8
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453#issuecomment-211859777>
>

bparees commented on pkg/build/controller/policy/policy.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// GetDefaultRunPolicies returns a set of all default run policies.
+func GetDefaultRunPolicies(lister buildclient.BuildLister, updater buildclient.BuildUpdater) []RunPolicy {
```
in what sense is this all default policies?  isn't it just all policies?

mfojtik replied:
fixed.

bparees commented on pkg/build/controller/policy/policy.go:
```diff
@@ -0,0 +1,116 @@
+// build configuration. It also returns the indication whether there are
+// currently running builds, to make sure there is no race-condition between
+// re-listing the builds.
+func GetNextConfigBuild(lister buildclient.BuildLister, namespace, buildConfigName string) (*buildapi.Build, bool, error) {
```
GetNextBuildForConfig ?

mfojtik replied:
Yeah, I was considering that one but I think it is long ;-)

bparees commented on pkg/build/controller/policy/policy.go:
```diff
@@ -0,0 +1,116 @@
+	now := unversioned.Now()
+	nextBuild.Status.StartTimestamp = &now
+	if err := updater.Update(nextBuild.Namespace, nextBuild); err != nil && !errors.IsConflict(err) {
+		return err
```
we need to be very careful we don't stall the queue by failing to update a build here.  the only thing that keeps processing going is update the build to cause the next event.

i think this is ok because the retry controller logic will retry the event that lead us down this path.

mfojtik replied:
right, I had a big comment here but I lose him somehow... I can add polling here to make sure the build is updated.

mfojtik replied:
i added wait.Poll() here to make sure the update happen.

bparees commented on pkg/build/controller/policy/policy.go:
```diff
@@ -0,0 +1,116 @@
+		case buildapi.BuildPhasePending, buildapi.BuildPhaseRunning:
+			hasRunningBuilds = true
+			return false
+		}
```
can't you also leave out any build that's in a completed/canceled/failed state?  in fact i think you should, otherwise you might update its start time later.

mfojtik replied:
@bparees line below this... i'm getting only "new" builds here... the hasRunningBuilds only indicates that there are running builds...

bparees replied:
that's what happens with late night reviews :(

bparees commented on pkg/build/controller/policy/serial_latest_only.go:
```diff
@@ -0,0 +1,92 @@
+	if len(bcName) == 0 {
+		return false, NewNoBuildConfigLabelError(build)
+	}
+	if err := kerrors.NewAggregate(s.cancelPreviousBuilds(build)); err != nil {
```
so any new SerialLatestOnly build that gets created will result in the cancellation of any existing queued builds, both Serial and SerialLatestOnly.

I think some comments in this code which outlines the various flows and outcomes depending on what's in the queue+already running and what's being added would be helpful.  Again i'm less concerned about the behavior in these edge cases, but i want to make sure we at least have it clearly defined.

mfojtik replied:
I will update the godoc. Yes, it will cancel all builds and it will not look at the policy those builds have. It can also cancel parallel builds if there are some in the "new" state for some reason.

mfojtik replied:
(updated the godoc for this function to mention the BC run policy change.

mfojtik commented:
@bparees updated, also added a check to not run any parallel build when there is already a serial build running.

Starting to write unit tests :-)

mfojtik commented:
added unit tests for policies.

mfojtik commented:
[test]

mfojtik commented:
@bparees added more extended tests...

@smarterclayton I think this is ready for final reviews (we have unit and extended test coverage, everything seems to operate as expected).

mfojtik commented:
[testextendedonly][extended:core(build configuration runPolicy)]

bparees commented on test/extended/builds/run_policy.go:
```diff
@@ -0,0 +1,222 @@
+				if counter == 2 {
+					break
+				}
+			}
```
don't you need to check this counter value somewhere?

mfojtik replied:
i don't think so :-) if this counter is not equal to 2, this test will timeout...

bparees replied:
hm.  that's not going to be a very nice/debuggable failure.  and what's the test timeout?  might be better to introduce your own shorter timeout.

mfojtik replied:
i reworked that test to check for the state of the builds... if they are completed we exit immediately with error... that means the only timeout we can hit here is when the build takes ages to become pending (which I don't think happens very often).

The counter is still there to know when to exit the loop, but it has no testing value.

mfojtik commented:
github request limit... [test]

kargakis commented on pkg/build/api/types.go:
```diff
@@ -559,10 +558,34 @@ type BuildConfigSpec struct {
+
+	// BuildRunPolicySerialLatestOnly schedules only the latest build to execute,
+	// cancelling all the previously queued build.
+	BuildRunPolicySerialLatestOnly BuildRunPolicy = "SerialLatestOnly"
```
SerialLatestOnly sounds too verbose. How about LatestOnly?

mfojtik replied:
already discussed... we are going with SerialLatestOnly for now... we can revisit for `v2` api.

jhadvig commented on pkg/build/api/types.go:
```diff
@@ -559,10 +558,34 @@ type BuildConfigSpec struct {
 
+	// RunPolicy describes how the new build created from this build
+	// configuration will be scheduled for execution.
+	// This is optional, if not specified we default to "parallel".
```
Just a small nit .. should be "Parallel"

mfojtik replied:
it is a good nit :-) we default to "Serial" not parallel.

bparees commented on test/extended/builds/run_policy.go:
```diff
@@ -0,0 +1,241 @@
+				if build.Name == startedBuilds[0] {
+					if buildutil.IsBuildComplete(build) {
+						break
+					}
```
you need at least one pre-check to ensure you went through the loop at least one time while build1 was pending.  Otherwise if build1 completes extremely fast, you'll never check the other 2 builds to ensure they were in pending.

something like:  if loopcount==0  && IsBuildComplete(build) -> fail test  (because now you have no way to confirm whether or not the other 2 builds would have been properly queued).

mfojtik replied:
@bparees fixed, I'm checking if the counter is equal to 2 at the end (so breaking the loop due to first build complete will not result in positive run).

smarterclayton commented on pkg/build/api/v1/types.go:
```diff
@@ -525,10 +525,34 @@ type BuildConfigSpec struct {
+	// RunPolicy describes how the new build created from this build
+	// configuration will be scheduled for execution.
+	// This is optional, if not specified we default to "Serial".
+	RunPolicy *BuildRunPolicy `json:"runPolicy,omitempty"`
```
Doesn't need to be a pointer

smarterclayton commented on test/extended/builds/run_policy.go:
```diff
@@ -0,0 +1,241 @@
+	kapi "k8s.io/kubernetes/pkg/api"
+)
+
+var _ = g.Describe("[builds][Slow] using build configuration runPolicy", func() {
```
I'd like at least one test of this to be "fast" and in conformance suite that verifies the behavior - a single test that creates the build, slates some very fast ones that don't push, and then generates multiples.  Basically, like deployment, this can and will flake, and I want a ~15-20s test in conformance that will generate flakes so we have to fix them.

mfojtik replied:
@smarterclayton each of this tests take ~20s to complete, but sure, I will add some 'fast' running build test. Also I have couple unit tests to verify the different run policy strategies on unit level to make sure we pass the build in correct order.

mfojtik replied:
@smarterclayton are you OK with doing the conformance tests as follow up? (there is already pretty good test coverage and we have follow up card for this work already in trello).

smarterclayton commented on test/extended/fixtures/run_policy/serial-latest-only-bc.json:
```diff
@@ -0,0 +1,42 @@
+{
```
Use yaml for fixtures from now on.

mfojtik commented:
@smarterclayton API updated and JSON converted to YAML.

mfojtik commented:
[test]

mfojtik commented:
[test]

mfojtik commented:
[test]

(integration flake)

mfojtik commented:
@smarterclayton @bparees merge while it is green :-)

jimmidyson commented:
@mfojtik Don't you make me rebase #7949 again!

bparees commented:
@mfojtik still needs api approved label from @smarterclayton

smarterclayton commented:
Approved, thanks for following process :)

bparees commented:
lgtm but we'll let @jimmidyson's PR merge first :)  (it just got tagged)

mfojtik commented:
@bparees rebased... will wait for tests to pass :-)

mfojtik commented:
[test]

mfojtik commented:
ok jenkins, [test] again.

mfojtik commented:
[merge]

(tests are flaking but nothing related to this PR)

smarterclayton commented:
Are all the flakes known?

On May 2, 2016, at 11:32 AM, Michal Fojtik <notifications@github.com> wrote:

[merge]

(tests are flaking but nothing related to this PR)

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8453#issuecomment-216268805>

mfojtik commented:
Yes as far I know (s2i quota, mysql passtrough)
On May 2, 2016 18:04, "Clayton Coleman" <notifications@github.com> wrote:

> Are all the flakes known?
>
> On May 2, 2016, at 11:32 AM, Michal Fojtik <notifications@github.com>
> wrote:
>
> [merge]
>
> (tests are flaking but nothing related to this PR)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453#issuecomment-216268805>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453#issuecomment-216277630>
>
